x/ecocredit/simulation/operations.go

```diff
@@ -0,0 +1,594 @@
+	TypeMsgRetire      = sdk.MsgTypeURL(&ecocredit.MsgRetire{})
+	TypeMsgCancel      = sdk.MsgTypeURL(&ecocredit.MsgCancel{})
+)
+
+// WeightedOperations returns all the operations from the module with their respective weights
+func WeightedOperations(
+	appParams simtypes.AppParams, cdc codec.JSONCodec,
+	ak ecocredit.AccountKeeper, bk ecocredit.BankKeeper, qryClient ecocredit.QueryClient) simulation.WeightedOperations {
+
+	var (
+		weightMsgCreateClass int
+		weightMsgCreateBatch int
+		weightMsgSend        int
+		weightMsgRetire      int
+		weightMsgCancel      int
+	)
+
+	appParams.GetOrGenerate(cdc, OpWeightMsgCreateClass, &weightMsgCreateClass, nil,
+		func(_ *rand.Rand) {
+			weightMsgCreateClass = WeightCreateClass
+		},
+	)
+
+	appParams.GetOrGenerate(cdc, OpWeightMsgCreateBatch, &weightMsgCreateBatch, nil,
+		func(_ *rand.Rand) {
+			weightMsgCreateBatch = WeightCreateBatch
+		},
+	)
+
+	appParams.GetOrGenerate(cdc, OpWeightMsgSend, &weightMsgSend, nil,
+		func(_ *rand.Rand) {
+			weightMsgSend = WeightSend
+		},
+	)
+
+	appParams.GetOrGenerate(cdc, OpWeightMsgRetire, &weightMsgRetire, nil,
+		func(_ *rand.Rand) {
+			weightMsgRetire = WeightRetire
+		},
+	)
+
+	appParams.GetOrGenerate(cdc, OpWeightMsgCancel, &weightMsgCancel, nil,
+		func(_ *rand.Rand) {
+			weightMsgCancel = WeightCancel
+		},
+	)
+
+	return simulation.WeightedOperations{
+		simulation.NewWeightedOperation(
+			weightMsgCreateClass,
+			SimulateMsgCreateClass(ak, bk, qryClient),
+		),
+		simulation.NewWeightedOperation(
+			weightMsgCreateBatch,
+			SimulateMsgCreateBatch(ak, bk, qryClient),
+		),
+		simulation.NewWeightedOperation(
+			weightMsgSend,
+			SimulateMsgSend(ak, bk, qryClient),
+		),
+		simulation.NewWeightedOperation(
+			weightMsgRetire,
+			SimulateMsgRetire(ak, bk, qryClient),
+		),
+		simulation.NewWeightedOperation(
+			weightMsgCancel,
+			SimulateMsgCancel(ak, bk, qryClient),
+		),
+	}
+}
+
+// SimulateMsgCreateClass generates a MsgCreateClass with random values.
+func SimulateMsgCreateClass(ak ecocredit.AccountKeeper, bk ecocredit.BankKeeper,
+	qryClient ecocredit.QueryClient) simtypes.Operation {
+	return func(
+		r *rand.Rand, app *baseapp.BaseApp, sdkCtx sdk.Context, accs []simtypes.Account, chainID string,
+	) (simtypes.OperationMsg, []simtypes.FutureOperation, error) {
+		admin := accs[0]
+		issuers := randomIssuers(r, accs)
+
+		ctx := regentypes.Context{Context: sdkCtx}
+		res, err := qryClient.Params(ctx, &ecocredit.QueryParamsRequest{})
+		if err != nil {
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgCreateClass, err.Error()), nil, err
+		}
+
+		params := res.Params
+		if params.AllowlistEnabled && !contains(params.AllowedClassCreators, admin.Address.String()) {
+			return simtypes.NoOpMsg(ecocredit.ModuleName, TypeMsgCreateClass, "not allowed to create credit class"), nil, nil // skip
+		}
+
+		spendable := bk.SpendableCoins(sdkCtx, admin.Address)
+		if spendable.IsAllLTE(params.CreditClassFee) {
```

Comment:
Are we sure we're generating cases where this is false, i.e. the admin has enough funds?

Reply:
Yes. Without this check, simulations are failing.

```
failed to execute message; message index: 0: 4stake is smaller than 32stake: insufficient funds [regen-network/cosmos-sdk@v0.43.0-regen-1/x/bank/keeper/send.go:187]
```